server/src/main/java/org/apache/kafka/server/ListOffsetsPartitionStatus.java

```diff
@@ -0,0 +1,122 @@
+    private volatile boolean completed = false;
+
+    private ListOffsetsPartitionStatus(
+        Optional<AsyncOffsetReadFutureHolder<FileRecordsOrError>> futureHolderOpt,
+        Optional<Long> lastFetchableOffset,
+        Optional<ApiException> maybeOffsetsError,
+        Optional<ListOffsetsPartitionResponse> responseOpt
+    ) {
+        this.futureHolderOpt = futureHolderOpt;
+        this.lastFetchableOffset = lastFetchableOffset;
+        this.maybeOffsetsError = maybeOffsetsError;
+        this.responseOpt = responseOpt;
+    }
+    
+    public static Builder builder() {
+        return new Builder();
+    }
+    
+    public static class Builder {
+        private Optional<AsyncOffsetReadFutureHolder<FileRecordsOrError>> futureHolderOpt = Optional.empty();
+        private Optional<Long> lastFetchableOffset = Optional.empty();
+        private Optional<ApiException> maybeOffsetsError = Optional.empty();
+        private volatile Optional<ListOffsetsPartitionResponse> responseOpt = Optional.empty();
+        
+        public Builder futureHolderOpt(Optional<AsyncOffsetReadFutureHolder<FileRecordsOrError>>  futureHolder) {
+            this.futureHolderOpt = futureHolder;
+            return this;
+        }
+        
+        public Builder lastFetchableOffset(Optional<Long> lastFetchableOffset) {
+            this.lastFetchableOffset = lastFetchableOffset;
+            return this;
+        }
+        
+        public Builder maybeOffsetsError(Optional<ApiException> maybeOffsetsError) {
+            this.maybeOffsetsError = maybeOffsetsError;
+            return this;
+        }
+        
+        public Builder responseOpt(Optional<ListOffsetsPartitionResponse> responseOpt) {
+            this.responseOpt = responseOpt;
+            return this;
+        }
+        
+        public ListOffsetsPartitionStatus build() {
+            return new ListOffsetsPartitionStatus(
+                    futureHolderOpt,
+                    lastFetchableOffset,
+                    maybeOffsetsError,
+                    responseOpt
+            );
+        }
+        
+    }
+
+    public Optional<AsyncOffsetReadFutureHolder<FileRecordsOrError>> futureHolderOpt() {
+        return futureHolderOpt;
+    }
+
+    public Optional<Long> lastFetchableOffset() {
+        return lastFetchableOffset;
+    }
+
+    public Optional<ApiException> maybeOffsetsError() {
+        return maybeOffsetsError;
+    }
+
+    public void responseOpt(Optional<ListOffsetsPartitionResponse> responseOpt) {
+        this.responseOpt = responseOpt;
+    }
+
+    public Optional<ListOffsetsPartitionResponse> responseOpt() {
+        return responseOpt;
+    }
+    
+    public void completed(boolean completed) {
+        this.completed = completed;
+    }
+
+    public boolean completed() {
+        return completed;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[responseOpt: %s, lastFetchableOffset: %s, " +
```

Comment:
`toString` without class-name looks odd. I understand that we ported the changes as is from scala -> java. And, the `toString` is used only in DelayedRemoteListOffsets logger, so should be fine.